clients/src/main/java/org/apache/kafka/common/protocol/Readable.java

```diff
@@ -26,34 +30,22 @@
     int readInt();
     long readLong();
     void readArray(byte[] arr);
+    int readUnsignedVarint();
 
-    /**
-     * Read a Kafka-delimited string from a byte buffer.  The UTF-8 string
-     * length is stored in a two-byte short.  If the length is negative, the
-     * string is null.
-     */
-    default String readNullableString() {
-        int length = readShort();
-        if (length < 0) {
-            return null;
-        }
+    default String readString(int length) {
         byte[] arr = new byte[length];
         readArray(arr);
         return new String(arr, StandardCharsets.UTF_8);
     }
 
-    /**
-     * Read a Kafka-delimited array from a byte buffer.  The array length is
-     * stored in a four-byte short.
-     */
-    default byte[] readNullableBytes() {
-        int length = readInt();
-        if (length < 0) {
-            return null;
+    default List<RawTaggedField> readRawTaggedField(List<RawTaggedField> unknowns, int tag, int size) {
+        if (unknowns == null) {
+            unknowns = new ArrayList<>();
         }
```

Comment:
Are we doing this to avoid the allocation of `unknowns` unless there is at least one unknown?

Reply:
yes